### PR TITLE
Refactor SnapshotsInProgress to Track Snapshots By Repository

### PR DESCRIPTION
**Pull request by original-brownbear**
First step in making `SnapshotsInProgress` easier to work with
by tracking snapshots per repository. This allows simplifing the
concurrency logic in a couple of places and sets up a follow-up
that would invert the current list of maps for snapshots that is
very hard to reason about in the concurrency logic into a map
of lists that maps repo-shard to snapshots to make the logic
more obviously correct.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:925]: I think this all still works if we just look at the one repo, right?

```suggestion
        if (snapshotsInProgress.forRepo(repository).size() == 1) {
```

(also extract that variable sooner)

That said, not sure this special-case optimisation really adds much after these changes, I'd be fine with just dropping it.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:946]: If `snapshots.isEmpty()` can we always just return `snapshotsInProgress.forRepo(repository)`?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1037]: Not directly related to this change, but could/should we check later snapshots too? We'd expect the first mention of every shard in the list to be active (or blocked by a delete) I think.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2994]: We don't actually consume these updates (we used to before introducing `SnapshotShardsUpdateContext` it seems) we just iterate over them. So maybe `updatesByRepo` would be a better name?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:3020]: This makes a new map for every repo with active updates. Often that's not many but I've certainly heard folks talking about having lots of repositories. Could we instead create a map of repository contexts (between the top-level update context and the individual entry contexts) and use it to do a one-off map creation instead?

**Review by DaveCTurner — COMMENTED** [review]
Overall looks good, some nice simplifications, and seems solid after merging with #78587. I left only small comments about local changes.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: ++

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: Yea did this and dropped the count == 1 optimization as well.

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java]: ++ good idea, went for this and used an iterator instead of the list now

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:62]: ++ done

File: server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java]: :facepalm:  Actually, this has always been a nonsensical assertion. Changed all spots to:

```
assertTrue(masterNode.clusterService.state().custom(SnapshotsInProgress.TYPE, SnapshotsInProgress.EMPTY).isEmpty());
```

... which is actually what this was supposed to assert (everything is done) instead of asserting that snapshots are either done or finalizing (which would be broken).

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java]: ++ done

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java]: Right, inlined and handled differently in the test

File: server/src/main/java/org/elasticsearch/snapshots/InFlightShardSnapshotStates.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/InFlightShardSnapshotStates.java]: Right, that was always weird, went with a constant and early breakout for the empty case now :)

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java]: ++ added

File: server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java:195]: This should be perfectly fine. I went through all the relevant code and there's only one place where this matters and that is the snapshot shards service. If the order were to actually change on a master failover, then the data nodes would re-inspect all the snapshots needlessly even though they didn't change. Other than that, there shouldn't be any situation where this has any impact as the master nodes start work from scratch in terms of local state anyway when things fail over to them and data nodes only interact with these objects via the snapshot shards service.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: Right, made this a little nicer :)

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2999]: I think I'd like to skip this one :) The planned follow-up here will redo the construction and updating of `SnapshotsInProgress` anyway to finally get away from all this magical building and updating of the entries list in just the right order so this is just going to be pretty short lived and good enough for that. Creating a few extra maps here, even if you have 10 repos should be completely irrelevant relative to the other work we're doing in this state update cost wise.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1015]: Yea there's a couple of other week points in this assertion that https://github.com/elastic/elasticsearch/pull/78587 revealed in fact. I think I'll be able to go for a set of way cleaner and stricter assertions in the follow-up to this one.

**Comment by original-brownbear** [comment]
Thanks David! All points addressed now I think :)

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:931]: nit: now we can just do this, no need for the intermediate builder and wrapper:

```suggestion
                    return List.of(entry);
```

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:935]: Likewise:

```suggestion
        return List.of();
```

**Review by DaveCTurner — APPROVED** [review]
LGTM (left one optional nit)

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:931]: I don't think we can, we have two nested loops here and we're only breaking the inner one.

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**
